### DISCUSSION TITLE
Model global weights are suddenly zero within local training code

### DISCUSSION THREAD

**Discussion by virginiafdez** [NVIDIA/NVFlare#3132] [Q&A] [answered]
### Python version (`python3 -V`)

3.8

### NVFlare version (`python3 -m pip list | grep "nvflare"`)

2.2.5

### Operating system

Ubuntu 22.04

### Have you successfully run any of the following examples?

- [X] hello-numpy-sag with simulator
- [X] hello-pt with simulator
- [ ] hello-numpy-sag with POC
- [ ] hello-pt with POC

### Please describe your question

I am relatively new to nvflare, and  Iam running a training code I need to trim up, and while running it on the nvflare simulator, I am seeing that the weights that are passed on from server to each site initially are set to zero.
The model starts with pre-loading weights when it is created. In the scatter and gather workflow, I am printing said weights by accessing the ._global_weights attribute of the fl_ctx object and they are fine (not zero).
But then, right after, the broadcast_and_wait function is called to launch the training task and if I go to the execute method of that task, which I imagine, is the first thing that gets executed, and print the fl_ctx weights, they are zero. 
Is there anything I should be aware of that might be causing this? Any function that is potentially zeroing or losing the weights that are correctly loaded in the server and passed on to the task?

Thanks a lot for any input!!

**Comment by virginiafdez** [comment]
Update:
I have noticed that SVT Privacy and Percentile Privacy are used. But if I print the output of the process_dxo method of said filters, the weights are still not zero, and yet when I get into the execute method of the Executor, the weights are all zero.
I have also noticed that if I remove the filters from the server JSON file, the weights stop being zero on the execute method of the trainer.
I cannot figure out, though, what is being called that is zeroing the weights.
I assume there is a check in the Controller for Filters, and if there are Filters, they are applied, and something else is happening which is the root cause of the problem.

**Comment by YuanTingHsieh** [comment]
@virginiafdez Thanks for you interest and question!

Could you switch to NVFlare 2.5 or the main branch and try again?
And can you please share the job folder that you are running?

**Comment by virginiafdez** [comment]
Hello!

I've managed to continue debugging the application; the weights are actually not zero, but they are still causing my training to NaN, something that does not happen when SVT is not used. I've narrowed it down to the impact that the privacy filters cause on the normalisation layers (because the NaNs only happen when eval() is used). 
I'm trying to look for a minimal impact of SVT on the pre-trained weights but I am not having any success; do you have any recommendations?
I'm currently using:   "fraction": 1,   "epsilon": 0.0001,  "noise_var": 1000000,  "gamma": 15
The weights that are getting filtered are those from a simple DenseNet121 that can be downloaded from the torch model zoo. 

About switching to NVFlare 2.5, I am unable to, as the production system we are using has 2.2.5 and cannot currently be updated.

**Reply by YuanTingHsieh** [reply]
@virginiafdez thanks for providing more info!

@holgerroth @ZiyueXu77 do you have good suggestion for using SVT? thanks!

**Reply by holgerroth** [reply]
This noise_var is super high. I would suggest SVT parameters more similar to this https://github.com/NVIDIA/NVFlare/blob/08a446efcd6db9789164a8cee8c0717835bebedc/examples/advanced/brats18/configs/brats_fedavg_dp/app/config/config_fed_client.json#L28

You could try some suggestions from this discussion https://github.com/NVIDIA/NVFlare/discussions/824#discussioncomment-3746368

**Reply by virginiafdez** [reply]
With the suggested results I get worse results somehow (nans). If I look at the filtered weights with these values (fraction 0.9, epsilon 0.001, noise_var 1.0, gamma 1e-4), most of then are zero: 

![filtered_0 001_1 0_0 0001_1 0](https://github.com/user-attachments/assets/9b904b73-a3a9-45ed-9954-74046969f8f5)
_Histogram of the flattened weights before and after going through the SVT filter_

As far as I have seen, the noise variance is dividing the scale of the noise that is added post-filtering (https://github.com/NVIDIA/NVFlare/blob/1be401cd9a7da6532b9282be068f242539231be3/nvflare/app_common/filters/svt_privacy.py#L110), so big values seem to lead to smaller added noise. 

If I remove the filtered running mean and variances of the densenet weights from the loading, the network collapses - because, I imagine, the filtering is much harsher and the network can't recover - but the validation results - at least - are not nan.

**Reply by holgerroth** [reply]
Interesting. @wyli might have some insights here.

Is your training working fine without the `SVTPrivacy` filter?

**Reply by holgerroth** [reply]
As a note, @virginiafdez , you could also start using the [GaussianPrivacy](https://github.com/NVIDIA/NVFlare/blob/f5cb4dd2fa20f6b7237aaec40d1a5fc80b4b5eea/research/quantifying-data-leakage/src/nvflare_gradinv/filters/gaussian_privacy.py#L25) filter from the gradient inversion example to get started. That one has less parameters that need to be tuned.

Again, the main thing is to make sure that your training works without filters and then we can go from there.

**Comment by virginiafdez** [comment, marked as answer]
More updates into this issue, and a summary of what happens in each scenario. I've been shuffling the following:

- With and without filters (SVT)
- Train in train() mode (obviously) and eval in train() or eval() mode
- Load specific weights during training

Regardless of SVT being applied or not, with all the weights (the ones coming out of SVT) loaded, the training results are fine. If I evaluate on train() mode, the results are fine as well.

When I use SVT, and I validate on eval() mode, as I should:
- nan happens if I load all the weights
- nan does not happen if I load all the weights except those coming out of the norm layers. I have filtered even more, and the issue seems to be coming out of the running_mean and running_variance values, which makes sense, cause they are the ones exhibiting a different behaviour on train and eval mode.

Since these values running mean and variances are not learnable, does it make sense for them to go through the filters anyway? I assume I could write a customised filter excluding specific keys and try it out. 
I am also surprised the behaviour from a Densenet121 (standard, pre-trained and taken from torch) is so dependent on these values. I have looked at the histograms of filtered and unfiltered values from the norm layers and although they do change, I don't find it so dramatic.  

![histogram_updated_weights](https://github.com/user-attachments/assets/aec2ec0a-aad8-44f3-83dd-db050b620104)

I was curious on what are the minimal-impact parameters you can set with SVT to avoid modifying the weights too much (although I know that the point of SVT is precisely to do something to the weights, it's useful for testing purposes when the filter has to be there but the parameters can be tweaked).

Besides this, the issue is clarified as this seems to be the root cause of the problem unless there is anything else coming from you that might point to a different direction.
